bldsys/cmake/shader_helper.cmake

```
@@ -0,0 +1,104 @@
+            set(STORED_SHADER_FILE_SPV "${CMAKE_SOURCE_DIR}/shaders/${SHADER_FILE_GLSL}.spv")
+
+            get_filename_component(SHADER_FILE_DIR ${SHADER_FILE_SPV} DIRECTORY)
+            file(MAKE_DIRECTORY ${SHADER_FILE_DIR})
+
+            add_custom_command(
+                OUTPUT ${SHADER_FILE_SPV}
+                COMMAND ${CMAKE_COMMAND} -E make_directory "${CMAKE_CURRENT_BINARY_DIR}"
+                COMMAND ${GLSLC_EXECUTABLE} -o "${SHADER_FILE_SPV}" --target-spv=spv${TARGET_SPIRV_VERSION} ${PROJECT_SOURCE_DIR}/shaders/${SHADER_FILE_GLSL}
+                COMMAND ${CMAKE_COMMAND} -E copy "${SHADER_FILE_SPV}" "${STORED_SHADER_FILE_SPV}"
+                COMMAND ${SPV_VALIDATOR_EXECUTABLE} "${SHADER_FILE_SPV}" || true
+                DEPENDS ${PROJECT_SOURCE_DIR}/shaders/${SHADER_FILE_GLSL}
+                COMMENT "Compiling ${SHADER_FILE_GLSL} to SPIR-V"
+                VERBATIM)
+
+            list(APPEND SHADER_FILES_SPV ${SHADER_FILE_SPV})
+        endforeach()
+    endif()
+
+    find_program(DXC_EXECUTABLE dxc HINTS $ENV{VULKAN_SDK}/bin)
+    if(NOT DXC_EXECUTABLE)
+        message(WARNING "dxc not found! Shaders with .hlsl extension will not be compiled to SPIR-V. Please set VULKAN_SDK to the Vulkan SDK path.")
+    endif()
+
+    # Compile HLSL shaders to SPIR-V
+    if (DXC_EXECUTABLE)
+        foreach(SHADER_FILE_HLSL ${TARGET_SHADER_FILES_HLSL})
+            # Skip header and OpenCL files
+            if(${SHADER_FILE_HLSL} MATCHES ".*\\.h$" OR ${SHADER_FILE_HLSL} MATCHES ".*\\.cl$")
+                continue()
+            endif()
+
+            set(SHADER_FILE_SPV "${SHADER_DIR}/${SHADER_FILE_HLSL}.spv")
+            set(STORED_SHADER_FILE_SPV "${CMAKE_SOURCE_DIR}/shaders/${SHADER_FILE_GLSL}.spv")
+
+            get_filename_component(SHADER_FILE_DIR ${SHADER_FILE_SPV} DIRECTORY)
+            file(MAKE_DIRECTORY ${SHADER_FILE_DIR})
+
+            get_hlsl_target_profile(${SHADER_FILE_HLSL} SHADER_PROFILE)
+
+            add_custom_command(
+                OUTPUT ${SHADER_FILE_SPV}
+                COMMAND ${CMAKE_COMMAND} -E make_directory "${CMAKE_CURRENT_BINARY_DIR}"
+                COMMAND ${DXC_EXECUTABLE} -spirv -fspv-target-env=vulkan1.3 -E main -T ${SHADER_PROFILE} -Fo "${SHADER_FILE_SPV}" ${PROJECT_SOURCE_DIR}/shaders/${SHADER_FILE_HLSL}
+                COMMAND ${CMAKE_COMMAND} -E copy "${SHADER_FILE_SPV}" "${STORED_SHADER_FILE_SPV}"
+                COMMAND ${SPV_VALIDATOR_EXECUTABLE} "${SHADER_FILE_SPV}" || true
+                DEPENDS ${PROJECT_SOURCE_DIR}/shaders/${SHADER_FILE_HLSL}
+                COMMENT "Compiling ${SHADER_FILE_HLSL} to SPIR-V"
+                VERBATIM)
+
+            list(APPEND SHADER_FILES_SPV ${SHADER_FILE_SPV})
+        endforeach()
+    endif()
+
+    add_custom_target(${TARGET_ID}_shaders DEPENDS ${SHADER_FILES_SPV})
+    add_dependencies(${TARGET_ID} ${TARGET_ID}_shaders)
+    add_dependencies(vkb__shaders ${TARGET_ID}_shaders)
```

I avoid `add_custom_target` because it has no mechanism to not run if the shaders already exist and are up to date.

Because all of the files in `SHADER_FILES_SPV` were marked as the output of `add_custom_command`, they should all automatically be marked as `GENERATED`, so I believe you cacan have them build automatically with the target, and only when out of date with the following change:

```suggestion
    target_sources(${TARGET_ID} PRIVATE ${SHADER_FILES_SPV})
```